verify_stream.go

```diff
@@ -92,12 +91,10 @@ func (v *verifyStream) read(p []byte) (n int, err error) {
 		// we have an error, we still want to return n, since
 		// those bytes are verified (by readBlock's
 		// post-condition).
-		if v.err != nil {
-			return n, v.err
-		}
+		return n, v.err
 	}
 
-	return n, nil
+	panic(fmt.Sprintf("Should never get here: state=%v", v.state))
```

Comment:
I wonder if the simplest structure would be:

```
if internal_buffer_empty and not_end_of_stream:
    read a block into internal buffer (short circuit on error)
    if last:
        assert end of stream (short circuit on error)
        set end of stream state
if internal_buffer_empty:
    return eof
else:
    return data from buffer
```

Could that cover all our cases without the need for the trailing panic or so many exit points?